### DISCUSSION TITLE
`HandlerMethodArgumentResolver` for `DataFetcher` methods

### DISCUSSION THREAD

**Discussion by filipegeric** [Netflix/dgs-framework#1180] [Ideas]
Hello all! :wave: 

It would be interesting to implement a feature similar to `HandlerMethodArgumentResolver` that exists in Spring WebFlux.
What would it look like:
```kotlin
class ExampleArgumentResolver : HandlerMethodArgumentResolver {
    override fun supportsParameter(parameter: MethodParameter): Boolean {
        parameter.hasParameterAnnotation(ExampleAnnotation::class.java)
    }

    override fun resolveArgument(dfe: DgsDataFetchingEnvironment): Any {
        val exampleContext = DgsContext.getCustomContext<ExampleContext>(dfe)
        return exampleContext.value
    }
}
```

And it would be used in the data fetcher class like this:

```kotlin
@DgsComponent
class ExampleDataFetcher {
    @DgsQuery(field = "example")
    fun fetchExample(arg: String, @ExampleAnnotation example: Int): String {
        return arg + " " + example.toString()
    }
}
```

How it can be done now:

```kotlin
@DgsComponent
class ExampleDataFetcher {
    @DgsQuery(field = "example")
    fun fetchExample(arg: String, dfe: DgsDataFetchingEnvironment): String {
        val exampleContext = DgsContext.getCustomContext<ExampleContext>(dfe)
        val example = exampleContext.value
        return arg + " " + example.toString()
    }
}
```

or:

```kotlin
@DgsComponent
class ExampleDataFetcher {
    @DgsQuery(field = "example")
    fun fetchExample(arg: String, dfe: DgsDataFetchingEnvironment): String {
        return arg + " " + dfe.getExampleValue().toString()
    }
}

fun DgsDataFetchingEnvironment.getExampleValue(): Int {
    val exampleContext = DgsContext.getCustomContext<ExampleContext>(this)
    return exampleContext.value
}
```

If something like this can be implemented I feel like it would be useful.
Let me know what you think, or if something like this is already possible.
I appreciate the feedback.
Thanks!

**Comment by srinivasankavitha** [comment]
Thanks for the suggestion. Actually that is exactly how the `@InputArgument` annotation works under the hood. However, we don't have an extension point for users to add their own argument resolvers at this point. This could be an interesting feature to add, I agree.

**Reply by srinivasankavitha** [reply]
Actually, I take that back. We do have a way to add this to data fetchers. Currently we do this: https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs-spring-boot-oss-autoconfigure/src/main/kotlin/com/netflix/graphql/dgs/autoconfig/DgsInputArgumentConfiguration.kt

You can implement a similar configuration class for beans of type `ArgumentResolver`. That is based on the HandlerMethodArgumentResolver mechanism, so should be what you are looking for.

**Reply by bystam** [reply]
I found this discussion thread because I am looking for the same thing.

I tried implementing my own `ArgumentResolver` and it worked perfectly. The only thing is that I noticed that the `ArgumentResolver` type is inside an "internal" package (`com.netflix.graphql.dgs.internal.method.ArgumentResolver`).

Is it safe to rely on this, or do you prefer us to avoid using this type?

**Reply by srinivasankavitha** [reply]
You're right. It is currently an internal interface, but I don't see why we couldn't make it more of a public feature. Will do that in an upcoming release. For now, please use the internal type.

**Reply by atengel** [reply]
Hey @srinivasankavitha! Just wanted to jump in here and add that having a public feature for this would be great for our use case too and wondering if there are any plans for this? Appreciate the workaround for now 🙂

**Reply by srinivasankavitha** [reply]
Yeah, I can make the change in an upcoming release soon. Probably in a couple weeks because we have a few priorities that need our attention at the moment.
